### PR TITLE
netty: netty tests resource leakage issue

### PR DESCRIPTION
**Pull request by vinodhabib**
I can see there are multiple tests in netty server and client side are randomly failing with resource leakage error and this PR will be fixing some of them at Server and Client side as below.

**Note** : Made some common method changes (as per Eric Suggestion/Reference) in all 3 tests files and currently tests which are shown below in Pending Server Base are one which are majorly impacting in both Server and Client test randomly which has common root cause while invoking the **dataFrame()** which we discussed earlier and tried suggestion as shown in the comment
https://github.com/grpc/grpc-java/issues/3353#issuecomment-2330796645

**Server Side UTs Fixes:**
sendFrameShouldSucceed()
inboundDataWithEndStreamShouldForwardToStreamListener()
inboundDataShouldForwardToStreamListener()
keepAliveEnforcer_sendingDataResetsCounters()
maxRstCount_withinLimit_succeeds()
maxRstCount_exceedsLimit_fails()

NettyHandlerTestBase.dataSizeSincePingAccumulates()

**Client Side UTs Fixes**
sendFrameShouldSucceed()

**Pending UTs at Server Base**
NettyHandlerTestBase.bdpPingWindowResizing
NettyHandlerTestBase.bdpPingLimitOutstanding
NettyHandlerTestBase.windowUpdateMatchesTarget
NettyHandlerTestBase.testPingBackoff

**Note** : you can refer the comment https://github.com/grpc/grpc-java/issues/3353#issuecomment-2295813721 for more details on these 4 pending Server Base UTs behavior

Fixes https://github.com/grpc/grpc-java/issues/3353

### CODE REVIEW

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java [hunk 1]
```diff
@@ -1346,7 +1368,7 @@ private ByteBuf emptyGrpcFrame(int streamId, boolean endStream) throws Exception
     try {
       return dataFrame(streamId, endStream, buf);
     } finally {
-      buf.release();
+      buf.clear();
```
Comment (kannanjgithub) [inline netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1371]: Revert this change.

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java [hunk 2]
```diff
@@ -106,18 +105,25 @@ protected void manualSetUp() throws Exception {}
...
+      channel().releaseOutbound();
+    }
+  }
+
   /**
    * Must be called by subclasses to initialize the handler and channel.
    */
   protected final void initChannel(Http2HeadersDecoder headersDecoder) throws Exception {
-    content = Unpooled.copiedBuffer("hello world", UTF_8);
+    //ByteBuf content = Unpooled.copiedBuffer("hello world", UTF_8);
```
Comment (kannanjgithub) [inline netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:126]: Remove commented line.

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java [hunk 3]
```diff
@@ -869,7 +873,8 @@ public void keepAliveEnforcer_sendingDataResetsCounters() throws Exception {
     future.get();
     for (int i = 0; i < 10; i++) {
       future = enqueue(
-          new SendGrpcFrameCommand(stream.transportState(), content().retainedSlice(), false));
+          //new SendGrpcFrameCommand(stream.transportState(), content().retainedSlice(), false));
```
Comment (kannanjgithub) [inline netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:876]: Remove commented line.

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java [hunk 4]
```diff
@@ -262,25 +268,27 @@ public void deliverFrame(
             if (frame != null) {
               ByteBuf bytebuf = ((NettyWritableBuffer) frame).bytebuf();
               compressionFrame.writeBytes(bytebuf);
+              bytebuf.release();
```
Comment (kannanjgithub) [inline netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:271]: As per comment https://github.com/grpc/grpc-java/issues/3353#issuecomment-2296725145 we should not be doing this in the test.

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java [hunk 5]
```diff
@@ -357,6 +365,7 @@ protected final ByteBuf captureWrite(ChannelHandlerContext ctx) {
     for (ByteBuf buf : captor.getAllValues()) {
       composite.addComponent(buf);
       composite.writerIndex(composite.writerIndex() + buf.readableBytes());
+      buf.clear();
```
Comment (kannanjgithub) [inline netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:368]: Revert.

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java [hunk 6]
```diff
@@ -1302,6 +1320,7 @@ public void maxRstCount_exceedsLimit_fails() throws Exception {
     manualSetUp();
     assertThrows(ClosedChannelException.class, () -> rapidReset(maxRstCount + 1));
     assertFalse(channel().isOpen());
+    channel().releaseOutbound();
```
Comment (kannanjgithub) [inline netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1323]: I don't see any outbound messages in the channel when this is called. Check here and in other places whether it is really necessary. 
Also since you also have it in _tearDown_ does it not suffice already?

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java (hunk 4 above)
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:271]: @kannanjgithub As I remember with this change I was able to fix the below issue with grpcFrame() method.
https://github.com/grpc/grpc-java/issues/3353#issuecomment-2296416975

As per my understanding the comment you mentioned above does not apply here as I am not modifying frame after passing it to channelRead(), also I am not doing anything here around the channelRead() method as mentioned by Eric as you mentioned above comment link. please confirm

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java (hunk 1 above)
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1371]: Fixed

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java (hunk 2 above)
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:126]: Fixed

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java (hunk 3 above)
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:876]: Fixed

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java (hunk 5 above)
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:368]: Fixed, it was part of trial and error.

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java (hunk 6 above)
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1322]: @kannanjgithub I can see even with tearDown() method the channel().releaseOutbound() entry is required as I have validated the multiple UT's without this outbound messages are not getting released. please find the below snippet for one of UT which says messages are not released without this entry and its printing the count of the same.

image

Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java]: @kannanjgithub As per our recent discussion, Removed the channel().releaseOutbound() entries in the UT's which are calling the rapidReset() method as its already handled here.

Also Reverted back the same in other UTs which is not required.

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java (hunk 1 above)
Comment (kannanjgithub) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1349]: As per documentation clear() does not release memory: https://github.com/netty/netty/blob/64c52255c073d4a308d3dbb56a12d3cca7692c47/buffer/src/main/java/io/netty/buffer/ByteBuf.java#L467

I tried with your PR and not calling either byteBuf.clear() or byteBuf.release() here and I did not observe the leak report for any of the 3 tests calling _emptyDataFrame()_ viz headersWithErrAndEndStreamReturnErrorButNotThrowNpe, streamErrorShouldNotCloseChannel and clientHalfCloseShouldForwardToStreamListener.

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java (hunk 4 above)
Comment (kannanjgithub) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:270]: What leak does this fix? I tried commenting this out and ran some of the tests that end up calling this method but I didn't observe any leak report.

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java (hunk 1 above)
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1349]: I think in the above comment and snippet we are discussing about 3 tests failures with refCnt decrement error if we use byteBuf.release()  instead of byteBuf.clear(), not about memory leakage issue

Comment (kannanjgithub) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1349]: No, my question is why byteBuf.release() is even required when it doesn't show any report leak.

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java (hunk 4 above)
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:270]: As per my knowledge the above change fixed the leakage issue in the grpcFrame() method as mentioned in the comment https://github.com/grpc/grpc-java/issues/3353#issuecomment-2295813721 also I remember this error will occur if you run the complete server test by 5+ times (sometimes I have seen these issue in 10 time also)

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java (hunk 1 above)
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1349]: you mean to say we need to remove this catch block which is doing the cleanup part, as this was an existing code?

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java [hunk 7]
```diff
@@ -425,14 +432,17 @@ public void windowUpdateMatchesTarget() throws Exception {
...
     int accumulator = wireSize;
     // 40 is arbitrary, any number large enough to trigger a window update would work
     for (int i = 0; i < 40; i++) {
       channelRead(grpcDataFrame(3, false, data));
+      channel().releaseOutbound();
       accumulator += wireSize;
     }
     long pingData = handler.flowControlPing().payload();
     channelRead(pingFrame(true, pingData));
+    channel().releaseOutbound();
```
[the remark this answers is not in the document]
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java]: Fixed

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java (hunk 1 above)
Comment (kannanjgithub) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1349]: No, leave it as is. For some reason I see 
- buf.release() and 
+ buf.clear()
making it appear like it was your change.

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java (hunk 4 above)
Comment (kannanjgithub) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:270]: The ByteBuf you release from the frame may be necessary but it should be done by calling frame.close() [here](https://github.com/grpc/grpc-java/blob/master/netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java#L271) and it will internally close the ByteBuf contained in the frame. Doing it in the Sink code that you did is not the meaningfule place to do this operation because the MessageFramer could technically be reused for multiple writes using the same ByteBuf in it even though the test is not doing that.

Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:270]: But in the below code, I can see the scope of the highlighted bytebuf only till writing it to compressionFrame after that no where it getting used.
correct me if i am wrong?

if (frame != null) {
              ByteBuf **bytebuf** = ((NettyWritableBuffer) frame).bytebuf();
              **compressionFrame**.writeBytes(**bytebuf**);
              **bytebuf**.release();
            }
            
the below leakage error still persist if I remove the bytebuf.release() here in grpcFrame()

image

Comment (kannanjgithub) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:270]: Was the above leak even after adding the call to frame.close() ?

**Review by kannanjgithub — APPROVED** [review]

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java [hunk 8]
```diff
@@ -262,25 +267,27 @@ public void deliverFrame(
...
+  protected final ByteBuf grpcDataFrame(int streamId, boolean endStream, byte[] content) {
+    return dataFrame(streamId, endStream, grpcFrame(content));
+  }
 
+  protected final ByteBuf dataFrame(int streamId, boolean endStream, ByteBuf content) {
     ChannelHandlerContext ctx = newMockContext();
-    new DefaultHttp2FrameWriter().writeData(ctx, streamId, content, 0, endStream, newPromise());
+    DefaultHttp2FrameWriter defaultHttp2FrameWriter = new DefaultHttp2FrameWriter();
+    defaultHttp2FrameWriter.writeData(ctx, streamId, content, 0, endStream, newPromise());
+    content.clear();
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:289]: What's the point of this `clear()`? Would anything go wrong if this line was just deleted?

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java (hunk 1 above)
Comment (ejona86) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1349]: I agree the `release()` looks like it should go away (the ownership is being transferred to `dataFrame()`, but `clear()` is useless here. Just delete the line?

Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1349]: Fixed

Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1349]: Fixed

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java [hunk 9]
```diff
@@ -16,8 +16,8 @@
 
 package io.grpc.netty;
 
+import static com.google.common.base.Charsets.UTF_8;
```
[the remark this answers is not in the document]
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:19]: As per your earlier suggestion/reference https://github.com/grpc/grpc-java/issues/3353#issuecomment-2265740153 to fix the leakage error this was changed

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java [hunk 10]
```diff
@@ -1315,7 +1313,9 @@ private void rapidReset(int burstSize) throws Exception {
     for (int period = 0; period < 3; period++) {
       for (int i = 0; i < burstSize; i++) {
         channelRead(headersFrame(streamId, headers));
+        channel().releaseOutbound();
```
[the remark this answers is not in the document]
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1316]: I have added these releaseOutbounds() here as i saw leakage error in the below 2 UTs earlier which are calling the rapidReset() method and it has outbound messages which are not released.

maxRstCount_withinLimit_succeeds()
maxRstCount_exceedsLimit_fails()

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java (hunk 8 above)
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:289]: Fixed. Was trying to fix the leakage issue in the dataFrame() as part of this modified the code but still the issue was not resolved. so reverting back to the original code.

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ejona86) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:19]: I don't understand. I said nothing about this, and it wasn't in the code change I linked.

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (ejona86) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java:1316]: `releaseOutbound()` is the sort of thing that only needs to be run once at the end of tests, to avoid leaks. Running it at the end of `@Before` could be okay if we don't want tests to see any queued messages, but there's no point to calling it multiple times here. Really, I'm surprised it isn't placed mostly just once in an `@After` to clean up after every test.

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:19]: Below snippet is captured from code change which you linked where I can see the UTF_8 import is already exists.

FileName - NettyHandlerTestBase.java

image

File: netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java]: Fixed

[the remark this answers is not in the document]
Comment (kannanjgithub) [reply netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java]: I had allowed that based on [this](https://github.com/grpc/grpc-java/blob/4e8f7df589cbe5a192e705ab372f2b27957bf662/netty/src/test/java/io/grpc/netty/NettyServerHandlerTest.java#L875) existing call inside a loop, which gave me the impression that the queue might be getting emptied without releasing and refilled multiple times during the test but that is not the case.

File: netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ejona86) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:19]: So you're agreeing with me?

The `\000` and such is significant. That is a gRPC message frame and a proto-encoded message.

[the remark this answers is not in the document]
Comment (kannanjgithub) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:19]: @vinodhabib What was the check style violation? You should fix that error and use " \000\000\000\000\015hello world".

[the remark this answers is not in the document]
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:19]: Fixed

[the remark this answers is not in the document]
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:19]: @kannanjgithub @shivaspeaks below is the checkstyle violation which faced during the build and removed the chars

image

[the remark this answers is not in the document]
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:19]: @kannanjgithub @shivaspeaks Any further updates here on how to proceed further?

[the remark this answers is not in the document]
Comment (ejona86) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:19]: If the checker is [following the style guide](https://google.github.io/styleguide/javaguide.html#s2.3.2-special-escape-sequences) (which is not a given; we've had to disable a lot of checks because they were broken), then it is complaining because `\015` can be encoded as `\r`.

Looking at `buildscripts/checkstyle.xml`, it looks like it is noticing the 015:
```
        <module name="IllegalTokenText">
            <property name="tokens" value="STRING_LITERAL, CHAR_LITERAL"/>
            <property name="format" value="\\u00(08|09|0(a|A)|0(c|C)|0(d|D)|22|27|5(C|c))|\\(0(10|11|12|14|15|42|47)|134)"/>
            <property name="message" value="Avoid using corresponding octal or Unicode escape."/>
        </module>
```

So replacing \015 with \r should fix it. But the worst-case here was to use `new byte[] {0, 0, 0, 0, 0x0D, ...}`, not to go and delete parts of the string.

[the remark this answers is not in the document]
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:19]: Fixed

[the remark this answers is not in the document]
Comment (ejona86) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:19]: This thread was started about the UTF_8 static import being changed. That hasn't been addresses.

[the remark this answers is not in the document]
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:19]: @ejona86 Request you elaborate more on this as it's not clear to me? as mentioned in the previous above comment (https://github.com/grpc/grpc-java/pull/11593#discussion_r1839658060) UTF_8 static import already exists and I used it.  I have not changed anything for the same.

not sure if you are asking to remove this UTF_8 static import and its usage? please confirm.

[the remark this answers is not in the document]
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java:19]: > java.nio.charset.StandardCharsets.UTF_8

Thanks for the clarification @ejona86

[the remark this answers is not in the document]
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java]: Fixed it and Ready for your Re-review.

[the remark this answers is not in the document]
Comment (vinodhabib) [reply netty/src/test/java/io/grpc/netty/NettyHandlerTestBase.java]: > @ejona86 Fixed it and Ready for your Re-review and Approval.

@ejona86 Gentle Remindar on your Re-review and Approval of this PR.

**Review by ejona86 — APPROVED** [review]
